### PR TITLE
8276662: Scalability bottleneck in SymbolTable::lookup_common()

### PR DESCRIPTION
**Pull request by dwhite-intel**
Symbol table lookup had an optimization added when the symbol table was split into a shared table (for CDS?) and a local table. The optimization tries to track which table successfully found a symbol, so it can try that table first the next time.

Symbol table lookup is used in many JVM operations, including classloading, serialization, and reflection.

At startup time, more symbols will be from the shared table, but over time lookup can will be from a mix of local and shared symbols (eg user classes still have java.lang.String fields or subclass from java.lang.Object), resulting in multiple threads fighting over the value of this global variable.

With enough threads and cores, this can result in "true sharing" cache line contention.

This fix solves the scalability issue by using a thread-local variable to track which table to search first instead of a static global. The "symbol table success history" also makes more sense logically as a thread local state than as a global state.

Other options would also solve the the scaling problem, but may change the behavior that we're trying to optimize, or add more overhead or complexity than warranted, such as:
- Statically preferring the shared or local table
- Checking the shared table first "early on". When enough local symbols have been added, then check the local table first.  
- Using a NUMA-aware set of N variables distributed over M threads.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8276662](https://bugs.openjdk.java.net/browse/JDK-8276662): Scalability bottleneck in SymbolTable::lookup_common()

### Reviewers
 * [Claes Redestad](https://openjdk.java.net/census#redestad) (@cl4es - **Reviewer**) ⚠️ Review applies to 7baf3fd4425820afefe5fa79e40dab134ec305c1
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to 7baf3fd4425820afefe5fa79e40dab134ec305c1
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**) ⚠️ Review applies to 7baf3fd4425820afefe5fa79e40dab134ec305c1
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to 7baf3fd4425820afefe5fa79e40dab134ec305c1

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6400/head:pull/6400` \
`$ git checkout pull/6400`

Update a local copy of the PR: \
`$ git checkout pull/6400` \
`$ git pull https://git.openjdk.java.net/jdk pull/6400/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6400`

View PR using the GUI difftool: \
`$ git pr show -t 6400`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6400.diff">https://git.openjdk.java.net/jdk/pull/6400.diff</a>

</details>

### CODE REVIEW

**Comment by shipilev** [comment]
I have been looking at this code in JDK-8260718 too. The switching heuristics needs some work, and some experimentation. I think for this sharing fix, slapping `THREAD_LOCAL` to `_lookup_shared_first` would be enough? This would throw away any questions about the changes to switching heuristics. We have one defined already, see e.g. `globalDefinitions_gcc.hpp`:

```
#define THREAD_LOCAL __thread
```

**Comment by shipilev** [comment]
> ```
> diff --git a/src/hotspot/share/classfile/symbolTable.cpp b/src/hotspot/share/classfile/symbolTable.cpp
> index fa966fc7b22..edf48a2c2c1 100644
> --- a/src/hotspot/share/classfile/symbolTable.cpp
> +++ b/src/hotspot/share/classfile/symbolTable.cpp
> @@ -93,3 +93,3 @@ static volatile bool   _has_items_to_clean = false;
>  static volatile bool _alt_hash = false;
> -static volatile bool _lookup_shared_first = false;
> +static THREAD_LOCAL bool _lookup_shared_first = false;
> ```

Woot, I think it works and even the simple "Hello World" startup seems to respond:

```
$ perf stat -r 1000 build/server-baseline/bin/java -Xms128m -Xmx128m Hello > /dev/null

 Performance counter stats for 'build/server-baseline/bin/java -Xms128m -Xmx128m Hello' (1000 runs):

             29.62 msec task-clock                #    1.165 CPUs utilized            ( +-  0.11% )
                93      context-switches          #    0.003 M/sec                    ( +-  0.14% )
                 0      cpu-migrations            #    0.010 K/sec                    ( +-  6.36% )
             2,608      page-faults               #    0.088 M/sec                    ( +-  0.01% )
        83,596,348      cycles                    #    2.822 GHz                      ( +-  0.22% )  (46.41%)
        10,204,346      stalled-cycles-frontend   #   12.21% frontend cycles idle     ( +-  0.40% )  (43.73%)
        28,164,763      stalled-cycles-backend    #   33.69% backend cycles idle      ( +-  0.36% )  (45.36%)
        63,232,465      instructions              #    0.76  insn per cycle         
                                                  #    0.45  stalled cycles per insn  ( +-  0.10% )  (53.59%)
        12,144,831      branches                  #  409.979 M/sec                    ( +-  0.11% )  (56.27%)
           448,789      branch-misses             #    3.70% of all branches          ( +-  0.25% )  (54.64%)

         0.0254298 +- 0.0000362 seconds time elapsed  ( +-  0.14% )

$ perf stat -r 1000 build/linux-x86_64-server-release/images/jdk/bin/java -Xms128m -Xmx128m Hello > /dev/null

 Performance counter stats for 'build/linux-x86_64-server-release/images/jdk/bin/java -Xms128m -Xmx128m Hello' (1000 runs):

             29.22 msec task-clock                #    1.166 CPUs utilized            ( +-  0.11% )
                93      context-switches          #    0.003 M/sec                    ( +-  0.14% )
                 0      cpu-migrations            #    0.009 K/sec                    ( +-  6.53% )
             2,605      page-faults               #    0.089 M/sec                    ( +-  0.01% )
        81,731,944      cycles                    #    2.797 GHz                      ( +-  0.24% )  (46.56%)
        10,163,520      stalled-cycles-frontend   #   12.44% frontend cycles idle     ( +-  0.40% )  (43.32%)
        26,563,080      stalled-cycles-backend    #   32.50% backend cycles idle      ( +-  0.36% )  (44.87%)
        63,660,354      instructions              #    0.78  insn per cycle         
                                                  #    0.42  stalled cycles per insn  ( +-  0.09% )  (53.44%)
        12,049,198      branches                  #  412.400 M/sec                    ( +-  0.12% )  (56.68%)
           447,857      branch-misses             #    3.72% of all branches          ( +-  0.22% )  (55.13%)

         0.0250630 +- 0.0000353 seconds time elapsed  ( +-  0.14% )
```

**Comment by shipilev** [comment]
> (Not that I know what USE_LIBRARY_BASED_TLS_ONLY is really for :-)

Some history: [JDK-8259372](https://bugs.openjdk.java.net/browse/JDK-8259372). Yes, it looks like generic code needs to be wrapped by `USE_LIBRARY_BASED_TLS_ONLY` for a time being. Like it is in `thread.cpp`, if we look for existing code.

**Comment by cl4es** [comment]
I agree the hint provided by `_lookup_shared_first` makes more sense as a thread local than global state, and that this particular flag seem a worthwhile THREAD_LOCAL candidate in face of potential footprint overhead.

**Review by cl4es — APPROVED** [review]
LGTM, although I think you need a review from a CDS maintainer. @iklam?

File: src/hotspot/share/classfile/symbolTable.cpp
Comment (dholmes-ora) [inline src/hotspot/share/classfile/symbolTable.cpp:98]: Can you add an explanatory comment as to why this is a thread-local please. Just a couple of lines. Thanks.

**Review by dholmes-ora — APPROVED** [review]
Based on the discussions this seems fine.

Thanks,
David

**Comment by iklam** [comment]
@dwhite-intel do you have any performance results that you can share? Also, the PR description needs to be updated to reflect the final version.

**Comment by iklam** [comment]
I did a quick check of the latest patch. Small start-up runs like this are susceptible to effects of dynamic frequency scaling due to CPU heat, so I interleaved the runs just to be safe:

```
for i in 1 2 3 4 5; do
    for v in old new; do
        echo $v.$i.txt
        perf stat -o $v.$i.txt -r 500 ./jdk_$v/bin/java \
             -Xmx128m -Xshare:on -version 2> /dev/null
    done
done


old.1.txt:          0.044077 +- 0.000425 seconds time elapsed  ( +-  0.96% )
old.2.txt:          0.043140 +- 0.000385 seconds time elapsed  ( +-  0.89% )
old.3.txt:          0.044393 +- 0.000415 seconds time elapsed  ( +-  0.93% )
old.4.txt:          0.043779 +- 0.000407 seconds time elapsed  ( +-  0.93% )
old.5.txt:          0.043221 +- 0.000390 seconds time elapsed  ( +-  0.90% )
geomean = 0.043719

new.1.txt:          0.043791 +- 0.000405 seconds time elapsed  ( +-  0.93% )
new.2.txt:          0.042716 +- 0.000364 seconds time elapsed  ( +-  0.85% )
new.3.txt:          0.043203 +- 0.000382 seconds time elapsed  ( +-  0.88% )
new.4.txt:          0.043437 +- 0.000398 seconds time elapsed  ( +-  0.92% )
new.5.txt:          0.043012 +- 0.000382 seconds time elapsed  ( +-  0.89% )
geomean = 0.043230
```

So at least on my machine (10 year old dual socket Xeon/Sandybridge) the patch doesn't seem to introduce any regression.

**Review by iklam — APPROVED** [review]

**Review by shipilev — APPROVED** [review]

**Comment by dwhite-intel** [comment]
@iklam, we've seen 5-15% improvement in serialization benchmarks on CascadeLake/IceLake systems, and almost 2x improvement on mutli-socket systems with more cores.

**Comment by dwhite-intel** [comment]
Thanks for the reviews everyone!
I believe all review comments have been responded to. If there's anything else, please let me know. Otherwise I plan on integrating by EOD.

**Pull request closed without merging**
